Show more items in the navbar

jessetrinity opened the pull request:
Currently the navigation bar in Visual Studio is not very useful for JavaScript files. We make the choice to show an item in the navbar based on the node type of its child items, and in some cases, its parent node.

By relaxing most the rules around what makes a navigation item "top level," we will get a more informative code hierarchy in the middle "object" dropdown, similar to how the navtree is presented in VSCode.

The right "member" dropdown always shows the child nodes of whichever node is selected in the middle dropdown, so we can keep navigation leaf nodes out of the middle dropdown to reduce clutter. 

Property assignments and class `this` properties will also be added to the navigation tree with this change.

Take the following sample:

```ts
const foo = {
    bar: {
        baz: {}
    },
    biz: function qux() {
        var quim;
    }
};

var someFunctions = function () {
    var a = () => { };
    var b = function () { };
    var c = function d() { var C; };
    var e = function () {
        var f = function () { var F; };
    };
};
```

**Before**

Property assignments do not show up, so we don't see any structure of nested objects (see `foo`) unless they have an 'important' node such as a function (`qux`). In those cases, the name of the function shows up in the child items list rather than the more intuitive member name.

![RoughBeforeFoo](https://user-images.githubusercontent.com/42591254/63951052-ef639580-ca31-11e9-9356-bbd1207ba3e6.png)

![RoughBeforeQux](https://user-images.githubusercontent.com/42591254/63951790-3aca7380-ca33-11e9-9665-a6576c319e57.png)

All of the child functions of `someFunctions` are deemed unimportant and only show in the member dropdown. Indeed even function `e` is woefully unimportant until its child `f` gets its own important child.

![BeforeSomeFunction](https://user-images.githubusercontent.com/42591254/63952276-1e7b0680-ca34-11e9-8d2e-a2d6577de0c4.png)

**After**

Property assignments now show up in navigation, and the name of the property shows instead of the function name.

We keep the decision to not show functions with empty bodies in the middle dropdown (though they do show as a child item in the right dropdown). `e` and `f` both show, as having _any_ child item is now considered a good enough reason to show up in navigation.

![RoughFooAfter](https://user-images.githubusercontent.com/42591254/63951807-428a1800-ca33-11e9-861d-1e7e5859e315.png)

jessetrinity commented:
@amcasey this is ready for a look when you get a chance.

amcasey commented on src/services/utilities.ts:
```diff
@@ -353,6 +353,7 @@ namespace ts {
             case SyntaxKind.MethodDeclaration:
             case SyntaxKind.MethodSignature:
                 return ScriptElementKind.memberFunctionElement;
+            case SyntaxKind.PropertyAssignment:
```
`ShorthandPropertyAssignment` and `SpreadAssignment` look like they might be related, but I don't know the specifics of either one.

amcasey replied:
`isVariableLike` shows a collection of related syntax kinds.

jessetrinity replied:
Currently spread assignments will just turn up the binding elements in the nav bar. e.g. `var [a, b, ...other]` will show `a`, `b`, and `...other`. If we want to do better I imagine we would have to resort to the sort of merging that is done elsewhere.

checker.ts has this comment on destructuring:
```ts
        //   let { x } = obj;  // BindingElement
        //   let [ x ] = obj;  // BindingElement
        //   { x } = obj;      // ShorthandPropertyAssignment
        //   { x: v } = obj;   // PropertyAssignment
        //   [ x ] = obj;      // Expression
```
Again we would show the `BindingElement` in `let { x } = obj;` as `x`, but I'm having trouble figuring out in what case a `ShorthandPropertyAssignment` node would show up in the ` { x } = obj;` example. I just see it as separate `Block` and `ExpressionStatement` nodes.

amcasey commented on src/services/navigationBar.ts:
```diff
@@ -716,6 +711,9 @@ namespace ts.NavigationBar {
         return topLevel;
 
         function isTopLevel(item: NavigationBarNode): boolean {
```
I would not object if "TopLevel" were renamed to something clearer. :wink:

amcasey commented on tests/cases/fourslash/navigationBarAnonymousClassAndFunctionExpressions.ts:
```diff
@@ -62,7 +62,7 @@ verify.navigationTree({
                         {
                             "text": "foo",
-                            "kind": "function"
+                            "kind": "property"
```
I'm a little surprised this changed.  You might want to ask someone with more JS experience than me whether they have a preference.

jessetrinity replied:
It should be `method` now (which appears to have the same icon as `function`).

jessetrinity commented:
> I saw three distinct changes:
> 
> 1. Property assignments go in the navtree
> 2. Anything with a child is "top-level" (vs special heuristics)
> 3. We use the property/variable name, rather than the initializer name, if the initializer is a function
> 
> Does that sound right? Did I miss anything?
> 
> This basically looks good - my main concern is that there might be other interesting SyntaxKinds.
> 
> I haven't read through the test changes yet, but I'm a little surprised to see that there aren't new tests. Maybe existing tests were expanded?

Your summary sounds correct. Mostly I just updated tests because this change should largely just make existing navbar tests more like the navtree ones.

I have added some more specific test for nested object literals to address (1) and a test for the function initializer `ScriptElementKind`.

jessetrinity commented:
> Was there a binding pattern test? I didn't see one.

There is an existing `BindingElement` test. I have updated it to check that a binding element with a function initializer is correctly named.

Also offline discussions have indicated that while people would prefer to see the `method` `ScriptElementKind` for properties with function initializers, there was less enthusiasm for doing the same for variables and binding elements. I have left those the same for now.

jessetrinity commented:
@mjbvz 

The changes here that would show up in VSCode are:

1. `PropertyAssignments`, `ShorthandPropertyAssignments`, and `SpreadAssignments` will show up in the navtree. Essentially all of
```ts
const a = {
    b: 0,
    c: function d(){},
    d,
    ....e
}
```
2. Where a variable declaration or a property assignment has a named function initializer, we use the name of the variable or property rather than the name of the function. In the above, we would expect to see `c` rather than `d` in navigation items.

amcasey commented on src/services/navigationBar.ts:
```diff
@@ -700,11 +703,11 @@ namespace ts.NavigationBar {
     /** Flattens the NavNode tree to a list, keeping only the top-level items. */
-    function topLevelItems(root: NavigationBarNode): NavigationBarNode[] {
-        const topLevel: NavigationBarNode[] = [];
+    function primaryNavBarItems(root: NavigationBarNode): NavigationBarNode[] {
```
Is the "primary" navbar the middle dropdown?

jessetrinity replied:
Explained it the comment where the selection actually occurs. I was trying to be 'editor agnostic' but I will just update 'primary' to 'middle' since (I think) this is VS only anyway.

amcasey replied:
What about something like "container"?
